plexpy/api.py

```diff
@@ -13,86 +16,134 @@
+        if kwargs.get('apikey', '') != plexpy.CONFIG.API_KEY:
+            self.msg = 'Incorrect API key'
 
-        if kwargs['apikey'] != plexpy.CONFIG.API_KEY:
-            self.data = 'Incorrect API key'
-            return
-        else:
-            self.apikey = kwargs.pop('apikey')
+        if 'out_type' in kwargs:
+            self.out_type = kwargs.pop('out_type')
+
+        if 'callback' in kwargs:
+            self.callback = kwargs.pop('callback')
+
+        if kwargs.get('apikey', '') == plexpy.CONFIG.API_KEY:
+            self.authenticated = True
+            del kwargs['apikey']
 
         if 'cmd' not in kwargs:
-            self.data = 'Missing parameter: cmd'
-            return
+            self.msg = 'Missing parameter: cmd. possible commands are: %s' % cmd_list
 
-        if kwargs['cmd'] not in cmd_list:
-            self.data = 'Unknown command: %s' % kwargs['cmd']
-            return
-        else:
+        if 'cmd' in kwargs and kwargs['cmd'] in cmd_list:
             self.cmd = kwargs.pop('cmd')
 
+        # Allow getApikey without the apikey
+        if self.cmd == 'getApikey':
+            self.authenticated = True
+            self.msg = None
+
+        if 'cmd' in kwargs and kwargs.get('cmd') not in cmd_list:
+            self.msg = 'Unknown command: %s possible commands are: %s' % (kwargs.get('cmd', ''), cmd_list)
+
         self.kwargs = kwargs
-        self.data = 'OK'
+
+    def _responds(self, result_type='success', data=None, msg=''):
+
+        if data is None:
+            data = {}
+        return {"response": {"result": result_type, "message": msg, "data": data}}
+
+    def _out_as(self, d):
+
+        if self.out_type == 'json':
+            cherrypy.response.headers['Content-Type'] = 'application/json;charset=UTF-8'
+            try:
+                out = json.dumps(d, indent=4, sort_keys=True)
+                if self.callback is not None:
+                    cherrypy.response.headers['Content-Type'] = 'application/javascript'
+                    # wrap with JSONP call if requested
+                    out = self.callback + '(' + out + ');'
+            # if we fail to generate the output fake an error
+            except Exception as e:
+                logger.info(u"API :: " + traceback.format_exc())
+                out = '{"response": {"result": "error", "data": "None", "message": "error while composing output: "' + str(e) + '"}}'
+        if self.out_type == 'xml':
+            cherrypy.response.headers['Content-Type'] = 'application/xml'
+            try:
+                out = xmltodict.unparse(d, pretty=True)
+            except ValueError as e:
+                logger.error('Failed to parse xml result')
+                try:
+                    out = xmltodict.unpase({"response": {"result": "error", "message": str(e), "data": None}}, pretty=True)
+                except Exception as e:
+                    logger.error('Failed to parse xml result error message')
+
+        return out
```

Comment:
`out` hasn't been initialised anywhere, going to throw an exception if none of the above conditions are true (or fail).